### ISSUE TITLE
String returned looks strange

### ISSUE THREAD

**Issue by AndreasReitberger** [3F/DllExport#190]
First, thank you for providing such an awsome library!
I already created my first function, however the returned string looks strange.

I need to rebuild this C function within C# using [Dllexport]

```
#define DllExport   __declspec( dllexport ) 

typedef enum {
  Error = -1,
  OK    = 0
} Result;

#ifdef __cplusplus
extern "C" {  // only need to export C interface if
  // used by C++ source code
#endif
#define DllExport   __declspec( dllexport ) 

DllExport Result FunctionCall(char* functionName, char* parameters,  char* returnString, int* returnValue);

#ifdef __cplusplus
}
#endif
```

This is what I created in C#

```
public enum Result
    {
        Error = -1,
        OK = 0,
    }
    public static class Dllhandler
    {

        [DllExport("FunctionCall", CallingConvention.Cdecl)]
        public static Result FunctionCall(string functionName, string parameters, string returnString, ref int returnValue)
        {
            try
            {
                switch (functionName.ToString())
                {
                    case "Combine":
                        returnString = "Reply from DLL";
                        returnValue = 10;
                        return Result.OK;
                    default:
                        return Result.Error;
                }
            }
            catch(Exception)
            {
                return Result.Error;
            }
        }
    }
```

While I get "10" from the int ReturnValue, the string ReturnString looks strange.

![image](https://user-images.githubusercontent.com/2952443/121882386-c26c9980-cd10-11eb-8cd9-15bb21157bac.png)

I guess I missed some encoding or do I need to Marshal the string?
Thanks for helping!

The question is related to:

* `DllExport -version`: 1.7.1
* Copy-paste from `Data` tab:

**Comment by 3F** [comment]
Hello,

1.7.1 is obsolete and contains critical [errors](https://github.com/3F/DllExport/blob/master/changelog.txt). Please use latest versions.

> I guess I missed some encoding 

Note C# strings/characters are 2 bytes based (UTF-16)

> or do I need to Marshal the string?

Marshall it correctly in any case. Pointers/addresses, or configure through attribute, etc.

For [Conari](https://github.com/3F/Conari) the calling of *FunctionCall(char\* functionName, char\* parameters,  char\* returnString, int\* returnValue)* may look like ~

```csharp
l.FunctionCall("fun", parameters, ref returnString, out returnValue);
```

Read more about data types [here](https://github.com/3F/DllExport/wiki/Quick-start#about-data-types)

> First, thank you for providing such an awsome library!

Thanks for using!

**Comment by AndreasReitberger** [comment]
Hi @3F
thanks for your reply. So far I tried everything to get the string returned right. 
Now I changed it to a byte[], however the results are the same.
Now I stuck.
The original C dll function does it like this:

```
#ifdef __cplusplus
extern "C" {  // only need to export C interface if
  // used by C++ source code
#endif
#define DllExport   __declspec( dllexport ) 

DllExport Result FunctionCall(char* functionName, char* parameters,  char* returnString, int* returnValue);

#ifdef __cplusplus
}
#endif
```

```
DllExport Result FunctionCall(char * functionName, char * parameters,  char * returnString, int * returnValue )
{
  char* fields[5];
  char str[1024] = "";//"a string of token,token...";
  int num = 0;
  bool result = false;
  returnString[0] = 0;//reset return string

  if (strcmp(functionName, "Combine") == 0)
  {
    if (num == 2)
    {
      String^ str0 = gcnew String(fields[0]);   //convert char array to string
      String^ str1 = gcnew String(fields[1]);   //convert char array to string
      String^ str2 =  pIctWrapper->Combine(str0, str1);
      *returnValue = str2->Length;
      sprintf_s(returnString, str2->Length+1, "%s", str2); //copy the result string to retrunString <<<====
      delete str0;                              //clean up
      delete str1;
      delete str2;
      result = true;
    }
    else
      result = false;
  }
  //return result
  if (result)
    return OK;
  else
    return Error;
}
```

And this is my equivalent in C#

```
[DllExport("FunctionCall", CallingConvention.Cdecl)]
public static Result FunctionCall([In] string functionName, [MarshalAs(UnmanagedType.LPTStr)] [In] string parameters, [In, Out] byte[] returnString, ref int returnValue)
{
    returnString = new byte[0];
    //Encoding encoding = Encoding.GetEncoding(28591);
    Encoding encoding = Encoding.GetEncoding(1252);
    String str = string.Empty;
    bool succeeded = false;
    try
    {
        switch (functionName.ToString())
        {
            case "Combine":
                str = "Reply from DLL";
                returnValue = 10;
                succeeded = true;
                break;
            default:
                break;
        }
    }
    catch(Exception)
    {
        str = "";
        return EXT_DLL_Result.EXT_DLL_Result_Error;
    }
    returnString = encoding.GetBytes(str);
    return succeeded ? Result.OK : Result.Error;
}
```

And this is what I get: ReturnString: ›Â}q
The caller is a linux program wrapped into a windows program and about 20 years old... 
I tried all common encodings, none seems to work. The default encoding for the text files, which the software is working with, are in ANSI.
Maybe you see what the problem might be..
Thank you for your time & help!

**Comment by 3F** [comment]
@AndreasReitberger I suggest [MSDN](https://www.msdn.microsoft.com) or any Dev Q&A such as [stackoverflow](https://stackoverflow.com/) to completely solve your task.

And no, [ref/out string support](https://github.com/3F/Conari/blob/ba1a82e1a7c9250e5806b5d72444be5c3b57850c/ConariTest/BindingContextTest.cs#L15) only using *Conari* (there you can also use something through *NativeString\<CharPtr\>* etc.)

For other case such for *MarshalAs* attribute, yes, *LPTStr* are correct for a 8-bit ansi but this will never work for `returnString = new byte[0];` You need also look at [StringBuilder](https://docs.microsoft.com/en-us/dotnet/api/system.text.stringbuilder) for a mutable strings if it's not *Conari.*

**tl;dr** You have completely different adaptation from original managed C++/CLI. Note that *returnString* does not change the received address at all, while you did. 

Also byte[] array must be marshaled in any case for unmanaged memory too because this points to allocated array in managed memory.

> The caller is a linux program wrapped into a windows program and about 20 years old...

A bit confused with your solution. Is this a managed (dotnet/CLR) wrapper too?

> I tried all common encodings, none seems to work. 

I suggest comparing bytes and addresses manually as I did [in past](https://www.youtube.com/watch?v=QXMj9-8XJnY) (somewhere, I don't remember timecodes). In general, some memory/heap inspection should help understand the any problem better and faster. But first of all, everything that I said above needs to be fixed.

**Comment by AndreasReitberger** [comment]
Well, changing to StringBuilder did the trick. Thank you so much for your help! :)
